src/sources/file.rs

```diff
@@ -193,6 +224,86 @@ pub fn file_source(
     }))
 }
 
+struct LineAgg<T> {
+    inner: T,
+    marker: Regex,
+    timeout: u64,
+    buffers: HashMap<String, Bytes>,
+    draining: Option<Vec<(Bytes, String)>>,
+    timeouts: DelayQueue<String>,
+}
+
+impl<T> LineAgg<T> {
+    fn new(inner: T, marker: Regex, timeout: u64) -> Self {
+        Self {
+            inner,
+            marker,
+            timeout,
+            draining: None,
+            buffers: HashMap::new(),
+            timeouts: DelayQueue::new(),
+        }
+    }
+}
+
+impl<T: Stream<Item = (Bytes, String), Error = ()>> Stream for LineAgg<T> {
+    type Item = (Bytes, String);
+    type Error = ();
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        loop {
+            if let Some(to_drain) = &mut self.draining {
+                if let Some((data, key)) = to_drain.pop() {
+                    return Ok(Async::Ready(Some((data, key))));
+                } else {
+                    return Ok(Async::Ready(None));
+                }
+            }
+
+            match self.inner.poll() {
+                Ok(Async::Ready(Some((line, src)))) => {
+                    // look for buffered content from same source
+                    if let Some(buffered) = self.buffers.remove(&src) {
+                        if self.marker.is_match(line.as_ref()) {
+                            // buffer the incoming line and flush the existing data
+                            self.buffers.insert(src.clone(), line);
+                            return Ok(Async::Ready(Some((buffered, src))));
+                        } else {
+                            // append new line to the buffered data
+                            let mut more = buffered.to_vec();
+                            more.push(b'\n');
+                            more.extend_from_slice(&line);
+                            self.buffers.insert(src, Bytes::from(more));
+                        }
+                    } else {
+                        // no existing data for this source so buffer it with timeout
+                        self.timeouts
+                            .insert(src.clone(), Duration::from_millis(self.timeout));
+                        self.buffers.insert(src, line);
+                    }
+                }
+                Ok(Async::Ready(None)) => {
+                    // start flushing all existing data, stop polling inner
+                    self.draining = Some(self.buffers.drain().map(|(k, v)| (v, k)).collect());
+                    continue;
+                }
+                Ok(Async::NotReady) => {
+                    // check for keys that have hit their timeout
+                    if let Ok(Async::Ready(Some(expired))) = self.timeouts.poll() {
```

Comment:
I'm curious to think that we probably want to call this poll before even attempting to poll the inner stream. Because I believe timers will not start until the first poll. I think I did something similar in partition batch sink thing where I poll the timers no matter what. This means if we are under heavy pressure a file may never get drained after a while if notready never gets returned from inner.poll.

Reply:
I had this a couple of different places and am definitely not sold on it staying here. The idea was to prioritize reading new data from the source and not worry too much if that means the timeout isn't perfectly strict.

A weird downside of moving it up and always polling is that we could theoretically flush data due to timeout when there is a new chunk of data ready that should have been appended. I don't think this would cause anyone problems in reality, but it seemed slightly cleaner to avoid it? There's also the fact that reading new data will cause old data to get flushed, while expiring old data doesn't cause new data to be read, so it seemed a little safer from a dataflow perspective to prioritize reading new data.

That said, I don't think it really matters much one way or the other and would be happy to move it up if you're more comfortable with that.

Reply:
You can always poll then do the actual timeout action in the notready section. The poll is cheap enough to do every poll of the stream.

Reply:
If you poll and it returns ready, you'd either have to do the timeout right then or stash it somewhere, right? Since you can only return a single event at a time.

Reply:
Yeah, so I would stash it in a `VecDequeue` then when you hit `NotReady` on the inner type just attempt to pull timeouts out of the queue. If its empty you know that no timer was ready. The key here is to start the timers you must poll them first iirc.

Reply:
Fine, fine, I did it even though I think it's dumb that I'm supposed to :smile: 

I did dig into the source a bit, and I'm pretty sure it's nothing like "starting the timers". It gets lazily registered with the actual timer wheel, but it shouldn't affect the actual deadline instant. Still, I didn't follow it far enough to see what happens if the registration doesn't happen until after the deadline has passed, and the "fix" is easy enough that I didn't mind too much.